src/state_causal_type.erl

```diff
@@ -0,0 +1,275 @@
+    orddict:is_empty(DataStore).
+
+%% @doc Get the Actor's next DotContext from DotCloud.
+-spec get_next_dot_context(dot_actor(), dot_cloud()) -> dot_context().
+get_next_dot_context(DotActor, DotCloud) ->
+    MaxCounter = ordsets:fold(fun({DotActor0, DotCounter0}, MaxValue0) ->
+                                      case (DotActor0 == DotActor) andalso
+                                              (DotCounter0 > MaxValue0) of
+                                          true ->
+                                              DotCounter0;
+                                          false ->
+                                              MaxValue0
+                                      end
+                              end, 0, DotCloud),
+    {DotActor, MaxCounter + 1}.
+
+%% @doc Insert a dot to DotCloud.
+-spec insert_dot_context(dot_context(), dot_cloud()) -> dot_cloud().
+insert_dot_context(DotContext, DotCloud) ->
+    ordsets:add_element(DotContext, DotCloud).
+
+%% @doc Merge two DotClouds.
+-spec merge_dot_clouds(dot_cloud(), dot_cloud()) -> dot_cloud().
+merge_dot_clouds(DotCloudA, DotCloudB) ->
+    ordsets:union(DotCloudA, DotCloudB).
+
+%% @doc Get SubDataStore pointed by the object from {dot_map, dot_map()}.
+-spec get_sub_data_store(term(), {{dot_map, data_store_type()}, dot_map()}) ->
+          {ok, data_store()}.
+get_sub_data_store(Object, {{dot_map, ValueDataStoreType}, DataStore}) ->
+    case orddict:find(Object, DataStore) of
+        {ok, SubDataStore} ->
+            {ok, SubDataStore};
+        error ->
+            {ok, new_data_store(ValueDataStoreType)}
+    end.
+
+%% @doc Insert (key: Object, value: SubDataStore) into {dot_map, dot_map()}.
+-spec insert_object(term(), data_store(), {{dot_map, data_store_type()}, dot_map()}) ->
+          {{dot_map, data_store_type()}, dot_map()}.
+insert_object(Object, SubDataStore, {{dot_map, ValueDataStoreType}, DataStore}) ->
+    {{dot_map, ValueDataStoreType}, orddict:store(Object, SubDataStore, DataStore)}.
+
+%% @doc Remove the Object from DataStore.
+%% The Object has to be in the DataStore. This can be checked by
+%% get_data_store().
+-spec remove_object(term(), {{dot_map, data_store_type()}, dot_map()}) ->
+          {{dot_map, data_store_type()}, dot_map()}.
+remove_object(Object, {{dot_map, ValueDataStoreType}, DataStore}) ->
+    {{dot_map, ValueDataStoreType}, orddict:erase(Object, DataStore)}.
+
+%% @doc Get all Objects from DataStore.
+-spec get_all_objects({{dot_map, data_store_type()}, dot_map()}) -> [term()].
+get_all_objects({{dot_map, _ValueDataStoreType}, DataStore}) ->
+    orddict:fetch_keys(DataStore).
+
+%% @doc Get the number of Objects from DataStore.
+-spec get_objects_count({{dot_map, data_store_type()}, dot_map()}) -> non_neg_integer().
+get_objects_count({{dot_map, _ValueDataStoreType}, DataStore}) ->
+    orddict:size(DataStore).
+
+%% @private
+%% @todo This function can be used for the compressing functionality.
```

Comment:
Can you please leave a comment at the top of the file so that's clear to other readers?